src/Protocol/Diaspora/Repository/DiasporaContact.php

```diff
@@ -0,0 +1,251 @@
+<?php
+/**
+ * @copyright Copyright (C) 2010-2022, the Friendica project
+ *
+ * @license GNU AGPL version 3 or any later version
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+namespace Friendica\Protocol\Diaspora\Repository;
+
+use Friendica\BaseRepository;
+use Friendica\Core\Logger;
+use Friendica\Core\Protocol;
+use Friendica\Core\System;
+use Friendica\Database\Database;
+use Friendica\Database\Definition\DbaDefinition;
+use Friendica\Model\APContact;
+use Friendica\Model\Contact;
+use Friendica\Model\Item;
+use Friendica\Model\ItemURI;
+use Friendica\Network\HTTPException;
+use Friendica\Network\Probe;
+use Friendica\Protocol\Diaspora\Entity;
+use Friendica\Protocol\Diaspora\Factory;
+use Friendica\Util\DateTimeFormat;
+use Psr\Http\Message\UriInterface;
+use Psr\Log\LoggerInterface;
+
+class DiasporaContact extends BaseRepository
+{
+	const ALWAYS_UPDATE                 = true;
+	const NEVER_UPDATE                  = false;
+	const UPDATE_IF_MISSING_OR_OUTDATED = null;
+
+	protected static $table_name = 'dcontact';
+
+	/**
+	 * @var Factory\DiasporaContact
+	 */
+	protected $factory;
+	/**
+	 * @var DbaDefinition
+	 */
+	private $definition;
+
+	public function __construct(DbaDefinition $definition, Database $database, LoggerInterface $logger, Factory\DiasporaContact $factory)
+	{
+		parent::__construct($database, $logger, $factory);
+
+		$this->definition = $definition;
+	}
+
+	/**
+	 * @param array $condition
+	 * @param array $params
+	 * @return Entity\DiasporaContact
+	 * @throws HTTPException\NotFoundException
+	 */
+	public function selectOne(array $condition, array $params = []): Entity\DiasporaContact
```

Comment:
Why is it called "selectOne"? Yes, it fetches a single row - but "One" is so totally unspecific. It could mean "select first", "select last" or "select random". I always read this "selectOne" like: "Pick one of matches, I don't care which one".

Reply:
`selectOne` means that we expect the condition to hit a primary or a unique key and so we return a single entity, not a collection. Behavior is dictated by the sorting in case the condition matches more rows than one.

Reply:
So we select the first entry in case of multiple ones ...

Reply:
Yes but the important part is the function, not the underlying implementation (which actually rests on `Database->selectFirst`). Functionally, this method returns one entity, as opposed to `select` which returns a collection of entities. This collection might have only one entity, but this sets expectations for their use. You would loop over a collection, while you wouldn't with a single entity.

Reply:
I still don't understand why we mustn't say "fetch the first contact with that url" instead of "fetch any contact with that url". I still replace "one" in my head automatically with "anyone". But yeah, I still don't grasp this whole class model. For me it's always like trying to obfuscate and to hide the functionality behind complicate models :-)

Reply:
It's actually the opposite, we hide the underlying technical layer under a higher functional layer. You're close to the database, so I expect you are thinking about database interactions for each app feature. It's good, but it's limiting. In Friendica, it ended up meaning arbitrary arrays everywhere, sometimes a field is set, sometimes not, who knows?

This new paradigm tries to install some sort of knowledge about what's passed around. What does `Database->selectFirst` returns? One database row? Maybe, and if does, from which table and with what fields? It depends! On the other hand, `Repository\DiasporaContact->selectOne` always returns one DiasporaContact entity (or throw an Exception if it doesn't exist), which properties are clearly defined, so you know what data you have on hand.

And yes, `Repository\DiasporaContact->selectOne` ends up using `Database->selectFirst` because it is useful as it is, but then we need to manufacture the actual entities we'll be using through the application using `Factory\DiasporaContact`.

Reply:
For anyone else interested, here's the comparison schema between the previous paradigm and the new paradigm:
[Entity-Factory-Repository.drawio.pdf](https://github.com/friendica/friendica/files/10008857/Entity-Factory-Repository.drawio.pdf)

Reply:
Still I think that "one" is a misleading word here. It's just like in a card trick when the magician says: "Select one card from the deck". Then you are expected to select a single card from anywhere in the deck. That is different to "select the first card from the deck".

Reply:
I get it, but in the case of `selectOne`, you're never supposed to have the whole deck. Really what you say when you use `selectOne` is "I want the king of clubs" and the magician hands you one card because only one can match, so it doesn't matter whether he picked the first card or the last card or any in the middle since the condition only matches one card.